File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        result = []
        if digits == "":
            return result

        n = len(digits)
        dic = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

        def dfs(i, letter):
            if i == n:
                result.append(letter)
                return

            for char in dic[ digits[i] ]:
                dfs(i+1, letter + char)
        
        dfs(0, "")
        return result

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> answer;
    map<char, vector<char>> dict = {
        {'2', {'a', 'b', 'c'}},
        {'3', {'d', 'e', 'f'}},
        {'4', {'g', 'h', 'i'}},
        {'5', {'j', 'k', 'l'}},
        {'6', {'m', 'n', 'o'}},
        {'7', {'p', 'q', 'r', 's'}},
        {'8', {'t', 'u', 'v'}},
        {'9', {'w', 'x', 'y', 'z'}}
    };

    void backtrack(string path, string digits, int idx, int n) {
        if (path.length() == n) {
            if (path == "") {
                return;
            }
            answer.push_back(path);
            return;
        }

        char key = digits[idx];

        for (auto c: dict[key]) {
            backtrack(path+c, digits, idx+1, n);
        }

        return;
    }

    vector<string> letterCombinations(string digits) {
        int n = digits.length();
        backtrack("", digits, 0, n);

        return answer;
    }
};